File: LC-comparism/LC solutions/LC_111 minDepth.py
# LC_111 minDepth
import collections
class TreeNode:
    def __init__(self,val = 0, left = None, right =None):
        self.val = val
        self.right = right
        self.left = left

class Solution:
    def minDepth(self, root:TreeNode)->int:

        if not root:
            return 0

        q = collections.deque()
        q.append((root, 1))                         # main difference here is we are passing a tuple 

        while q:
            node, depth = q.popleft()
                                                    # we could see no for loop here like traversing level
            if not node.left and not node.right:    
                return depth                        # where the loop ends, return depth not break

            if node.left:
                q.qppend((node.left, depth + 1))    # passing two arguments at the same time

            if node.right:
                q.qppend((node.right, depth + 1))


#-------------------------- Standard BFS approach ---------------------
def minDepth(self, root:TreeNode):
    if not root:
        return None
    depth = 1
    q = [root]

    while q:
        for i in range(len(q)):
            node = q.pop(0)
            if node.left is None and node.right is None:
                return depth                        # all implementation is return depth not break
            
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        depth += 1                                  # depth plus one in python(behind if) is different in java(before if)
    return depth                                    # for loops ends by the first node 3 then depth initial changed from 1 to 2
            



#-------------------------- BFS ---------------------------------

def minDepth(self, root:TreeNode):
    to_visit = collections.deque([(root,1)])

    while to_visit:
        node, ct = to_visit.popleft()
        if not node:
            continue
        if not node.left and not node.right:
            return ct
        to_visit += [(node.left, ct + 1),(node.right, ct + 1)]

    return 0


#---------------------------BFS without input a tuple ------------

def minDepth(self, root:TreeNode):
    if not root:
        return 0
    stack = [root]
    level = 1
    while stack:
        for node in stack:                                  # condition judgement putting here
            if not node.left and not node.right:            
                return level                                # without setting break, when condition met, while loop stop
        level += 1
        stack = [child for node in stack for child in (node.left, node.right) if child]

File: LC-comparism/LC solutions/LC_111 mindepth.java
// LC_111 Minimum Depth of Binary Tree
class Solution{
    public int minDepth(TreeNode root) {
        int depth = 0;
        if(root == null) 
        return depth;
    }
    Queue<TreeNode> queue = new ArrayDeque<TreeNode>();
    queue.offer(root);
    loop:while(!queue.isEmpty()){
        int size = queue.size();
        for(int i = 0; i < size; i++){
            TreeNode node = queue.poll();
            if(node.left == null && node.right == null){
                break loop;
            }
            if(node.left != null)
                queue.offer(node.left);
            if(node.right != null)
                queue.offer(node.right);
        }    
    }
    return depth;
}

// ------------------------- other approach -------------------------

class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        // root 本身就是一层，depth 初始化为 1
        int depth = 1;

        while (!q.isEmpty()) {
            int sz = q.size();
            /* 遍历当前层的节点 */
            for (int i = 0; i < sz; i++) {
                TreeNode cur = q.poll();
                /* 判断是否到达叶子结点 */
                if (cur.left == null && cur.right == null)
                    return depth;
                /* 将下一层节点加入队列 */
                if (cur.left != null)
                    q.offer(cur.left);
                if (cur.right != null)
                    q.offer(cur.right);
            }
            /* 这里增加步数 */
            depth++;
        }
        return depth;
    }
}
// 详细解析参见：
// https://labuladong.github.io/article/?qno=111
